### PR TITLE
queries

### PR DESCRIPTION
**Pull request by cendhu**
Signed-off-by: senthil <cendhu@gmail.com>

### CODE REVIEW

File: docs/getting-started/transactions/gosdk/datatx.md [hunk 1]
```diff
@@ -507,100 +474,39 @@ func deleteKey1(tx bcdb.DataTxContext) error {
...
+For updating `key2`, first, we fetch `key2` by calling `tx.Get("db2", "key2")`. Next, the fetched value is unmarshalled
+into a person object such that fields such as `age` and `graduated` can be updated. Then, the updated value is
+stored by calling `tx.Put("db2", "key2", jVal, m.AccessControl)`.
 
-func checkExistenceOfKey3(tx bcdb.DataTxContext) error {
-	v, m, err := tx.Get("db2", "key3")
-	if err != nil {
-		return err
-	}
+Finally, the transaction is committed by calling `dbtx.Commit(true)`. The argument true denotes that the synchronous submission. As a result, the `Commit()`
```
Comment (gennadylaventman) [inline docs/getting-started/transactions/gosdk/datatx.md:495]: `tx.Commit(true)`

File: docs/getting-started/transactions/gosdk/datatx.md [hunk 2]
```diff
@@ -305,45 +288,26 @@ func main() {
...
-	"fmt"
-)
+To delete the key `key1`, we need to call `tx.Delete("db2", "key1")`. Note that this is a blind delete. If you want to avoid the
+blind delete, we need to first fetch `key1` by calling `tx.Get("db2", "key1")` but ignoring the returned value and version. Next,
+we need to call `tx.Delete("db2", "key1")` to delete the key `key1`.
 
-func main() {
-	db, err := createConnection()
-    // if err is not nil, print and return
+Finally, the transaction is committed by calling `dbtx.Commit(true)`. The argument true denotes that the synchronous submission. As a result, the `Commit()`
```
Comment (gennadylaventman) [inline docs/getting-started/transactions/gosdk/datatx.md:302]: `tx.Commit(true)`

File: docs/getting-started/transactions/gosdk/usertx.md [hunk 3]
```diff
@@ -179,13 +172,13 @@ In the above code, we have created the user `alice` with read permission to the
 we have created the user `bob` with only read permission to both `db1` and `db2`.
 
 Once the users are created, `tx.PutUser("alice", nil)` stores the user `alice` to the database while `tx.PutUser("bob", bobACL)` stores the user `bob`
-to the database with read-write ACL on the user `bob`. 
+to the database with read-write ACL on the user `bob`.
```
Comment (gennadylaventman) [inline docs/getting-started/transactions/gosdk/usertx.md:175]: `tx.PutUser("alice", nil)` ?!?!?!?

**Review by gennadylaventman — COMMENTED** [review]
Some comments

File: docs/getting-started/transactions/gosdk/usertx.md (hunk 3 above)
Comment (cendhu) [reply docs/getting-started/transactions/gosdk/usertx.md:175]: I don't understand

File: docs/getting-started/transactions/gosdk/datatx.md (hunk 2 above)
Comment (cendhu) [reply docs/getting-started/transactions/gosdk/datatx.md]: Done

File: docs/getting-started/transactions/gosdk/datatx.md (hunk 1 above)
Comment (cendhu) [reply docs/getting-started/transactions/gosdk/datatx.md:503]: Done

File: docs/getting-started/transactions/gosdk/datatx.md [hunk 4]
```diff
@@ -113,62 +114,69 @@ func main() {
...
+
+const (
+	AccessControl_ANY AccessControlWritePolicy = 0
+	AccessControl_ALL AccessControlWritePolicy = 1
 )
+```
+The ACL holds
+ 1. `ReadUsers`: a list of users who can only read the key
+ 2. `ReadWriteUsers`: a list of users who can read and write the key.
+ 3. `SignPolicyForWrite`: denotes whether signature of all users in the `ReadWriteUsers` is needed to perform writes to the key.
```
[the remark this answers is not in the document]
Comment (cendhu) [reply docs/getting-started/transactions/gosdk/datatx.md:162]: Done. Anyway, this should be written as part of the Architecture and Design. We have a section for the ACL there.

File: docs/getting-started/queries/gosdk/user.md [hunk 5]
```diff
@@ -3,9 +3,34 @@ id: user
 title: Query an User Information
 ---
 
-### Source Code
+To query credential associated with an user, for now, we need to use a user administration transaction context.
+The `GetUser()` method in the `UsersTxContext` retrieve a given key from a given database.
```
[the remark this answers is not in the document]
Comment (cendhu) [reply docs/getting-started/queries/gosdk/user.md]: Done

File: docs/getting-started/queries/gosdk/simple-data-query.md [hunk 6]
```diff
@@ -1,8 +1,32 @@
...
+:::info pre-requisite
+For the example shown here to work, we need to have
+
+ 1. Two databases named `db1` and `db2` in the orion-server. If you have not created these two databases,
+refer to [creates databases using SDK](../../transactions/gosdk/dbtx#1-creation-of-databases) to create `db1` and `db2`.
+ 2. Two users named `alice` and `bob`. If you have not created these users already, refer to [create users using SDK](../../transactions/gosdk/usertx#1-addition-of-users).
+
+Finally, [Create a connection](../../pre-requisite/gosdk#creating-a-connection-to-the-orion-cluster) and
+[Open a database session](../../pre-requisite/gosdk#opening-a-database-session).
+:::
```
[the remark this answers is not in the document]
Comment (cendhu) [reply docs/getting-started/queries/gosdk/simple-data-query.md:30]: Done

File: docs/getting-started/queries/gosdk/db.md [hunk 7]
```diff
@@ -45,16 +50,14 @@ func main() {
...
 ```
 
-### Source Code Commentry
+## Source Code Commentry
 For simplicity, not all errors are handled in this code. Further, the implementation of `createConnection()` and `openSession()` can be found [here](../../pre-requisite/gosdk).
 
-The `session.DBsTx()` starts a new database administration transaction and returns the database administration transaction context. We can then perform all
-administrative activities using the transaction context.
+The `session.DBsTx()` starts a new database administration transaction and returns the database administration transaction context. Note that we use this context
+only for the query purpose and not perform any database administrative operation.
```
[the remark this answers is not in the document]
Comment (cendhu) [reply docs/getting-started/queries/gosdk/db.md]: Done

File: docs/getting-started/queries/gosdk/db.md [hunk 8]
```diff
@@ -3,6 +3,9 @@ id: db
 title: Check the Existence of a Database
 ---
 
+To check the existence of a database, for now, we need to use a database context. The `Exists()` of method in
```
[the remark this answers is not in the document]
Comment (cendhu) [reply docs/getting-started/queries/gosdk/db.md]: Done

File: docs/getting-started/queries/gosdk/complex-data-query.md [hunk 9]
```diff
@@ -1,5 +1,1311 @@
...
+	// if err is not nil, print and return
+
+	q, err := session.JSONQuery()
+	// if err is not nil, print and return
+
+	queryStr := `
+	{
+		"selector": {
+			"name": {
+				"$gt": "e",
```
[the remark this answers is not in the document]
Comment (cendhu) [reply docs/getting-started/queries/gosdk/complex-data-query.md:937]: In the subsequent PR I will explain what is the default behaviour

File: docs/getting-started/queries/gosdk/complex-data-query.md [hunk 10]
```diff
@@ -1,5 +1,1311 @@
...
+
+	session, err := openSession(db, "alice")
+	// if err is not nil, print and return
+
+	q, err := session.JSONQuery()
+	// if err is not nil, print and return
+
+	queryStr := `
+	{
+		"selector": {
```
[the remark this answers is not in the document]
Comment (cendhu) [reply docs/getting-started/queries/gosdk/complex-data-query.md:415]: In the subsequent PR.

File: docs/getting-started/queries/gosdk/complex-data-query.md [hunk 11]
```diff
@@ -1,5 +1,1311 @@
...
+
+person5: {"name":"e","age":31,"graduated":true,"rating":0,"degree":"master"}
+person6: {"name":"f","age":24,"graduated":true,"rating":12,"degree":"bachelor"}
+person8: {"name":"h","age":38,"graduated":true,"rating":1230,"degree":"master"}
+person3: {"name":"c","age":20,"graduated":false,"rating":-4}
+```
+
+</details>
+
+## 3) Greater than (`$gt`) Operator
```
[the remark this answers is not in the document]
Comment (cendhu) [reply docs/getting-started/queries/gosdk/complex-data-query.md:393]: In the subsequent PR.

File: docs/getting-started/queries/gosdk/complex-data-query.md [hunk 12]
```diff
@@ -1,5 +1,1311 @@
...
+                }
+
+        }
+
+person4: {"name":"d","age":19,"graduated":false,"rating":-1}
+```
+
+</details>
+
+## 2) Not Equal Operator (`$neq`)
```
[the remark this answers is not in the document]
Comment (cendhu) [reply docs/getting-started/queries/gosdk/complex-data-query.md:263]: In the subsequent PR.

File: docs/getting-started/queries/gosdk/complex-data-query.md [hunk 13]
```diff
@@ -1,5 +1,1311 @@
...
+	// if err is not nil, print and return
+
+	fmt.Println("transaction with txID " + txID + " got committed in the block " + strconv.Itoa(int(receipt.GetHeader().GetBaseHeader().GetNumber())))
+	return nil
+}
+```
+
+</details>
+
+## 1) Equal Operator (`$eq`)
```
[the remark this answers is not in the document]
Comment (cendhu) [reply docs/getting-started/queries/gosdk/complex-data-query.md:141]: In the subsequent PR.

File: docs/getting-started/queries/gosdk/complex-data-query.md [hunk 14]
```diff
@@ -1,5 +1,1311 @@
 ---
 id: complex-data-query
-title: Query Data Using JSON Fields
+title: Query Data Using Fields in JSON value
 ---
 
+Once a database session is created, we can call `session.JSONQuery()` to get the implementation of `Execute()` method.
```
[the remark this answers is not in the document]
Comment (cendhu) [reply docs/getting-started/queries/gosdk/complex-data-query.md]: Done.

File: docs/getting-started/queries/gosdk/complex-data-query.md [hunk 15]
```diff
@@ -1,5 +1,1312 @@
 ---
 id: complex-data-query
-title: Query Data Using JSON Fields
+title: Query Data Using Fields in JSON value
```
Comment (tock-ibm) [inline docs/getting-started/queries/gosdk/complex-data-query.md:3]: in a JSON Object

File: docs/getting-started/queries/gosdk/complex-data-query.md [hunk 16]
```diff
@@ -1,5 +1,1312 @@
...
 ---
 
+Once a database session is created, we can call `session.JSONQuery()` to get `JSONQuery` object.
+
```
Comment (tock-ibm) [inline docs/getting-started/queries/gosdk/complex-data-query.md:7]: to get a

File: docs/getting-started/queries/gosdk/complex-data-query.md [hunk 17]
```diff
@@ -1,5 +1,1312 @@
...
+// 			}
+// 		}
+//   }
+// }
+type JSONQuery interface {
+	Execute(dbName, query string) ([]*types.KVWithMetadata, error)
+}
+```
+
+To query a data, we need to call `Execute()` method of the object from above.
```
Comment (tock-ibm) [inline docs/getting-started/queries/gosdk/complex-data-query.md:36]: To query data 
-> data is plural, datum is singular, to query a record to query a value but to query data

this error repeats in several places

File: docs/getting-started/queries/gosdk/db.md [hunk 18]
```diff
@@ -3,6 +3,9 @@ id: db
 title: Check the Existence of a Database
 ---
 
+To check the existence of a database, for now, we need to use a database context. The method `Exists()`
```
Comment (tock-ibm) [inline docs/getting-started/queries/gosdk/db.md:6]: why the "for now"... I know we have plans to improve it, but this does not have to be reflected in this text

File: docs/getting-started/queries/gosdk/simple-data-query.md [hunk 19]
```diff
@@ -1,8 +1,33 @@
 ---
 id: simple-data-query
-title: Query Data Using Keys
+title: Query a Data Using Keys
```
Comment (tock-ibm) [inline docs/getting-started/queries/gosdk/simple-data-query.md:3]: To query data 
(plural) 
To query records
to query a record

File: docs/getting-started/queries/gosdk/simple-data-query.md [hunk 20]
```diff
@@ -1,8 +1,33 @@
...
 ---
 
+To query a data using the key, for now, we need to use a database context. The `Get()` method in the `DataTxContext`
```
Comment (tock-ibm) [inline docs/getting-started/queries/gosdk/simple-data-query.md:6]: query data

File: docs/getting-started/queries/gosdk/user.md [hunk 21]
```diff
@@ -3,9 +3,34 @@ id: user
 title: Query an User Information
 ---
```
Comment (tock-ibm) [inline docs/getting-started/queries/gosdk/user.md:4]: query user information

File: docs/getting-started/queries/gosdk/user.md [hunk 22]
```diff
@@ -3,9 +3,34 @@ id: user
...
 
-### Source Code
+To query credential associated with an user, for now, we need to use a user administration transaction context.
```
Comment (tock-ibm) [inline docs/getting-started/queries/gosdk/user.md:6]: a credential, a user

Comment (tock-ibm) [reply docs/getting-started/queries/gosdk/user.md:6]: for now? I suggest get rid of all those

File: docs/getting-started/transactions/curl/datatx.md [hunk 23]
```diff
@@ -207,9 +210,9 @@ use the `base64_decoder` utility.
 The result contains the base64 decoded `value` associated with the key and also the `access_control` and `version` as part of the
 `metadata`. In order to verify the signature, do not use the output of `base64_decoder` utility but the `cURL` output.
 
-## (3) Updation of an existing state
+## 2) Updation of an existing state
```
Comment (tock-ibm) [inline docs/getting-started/transactions/curl/datatx.md:213]: Updation is only valid India... "Updating an existing state"

File: docs/getting-started/transactions/curl/datatx.md [hunk 24]
```diff
@@ -490,13 +493,13 @@ curl \
 }
 ```
 
-## (5) Creation, Updation, Deletion of states within a single transaction
+## 4) Creation, Updation, Deletion of states within a single transaction
```
Comment (tock-ibm) [inline docs/getting-started/transactions/curl/datatx.md:497]: Updation - replace

**Comment by tock-ibm** [comment]
I did a quick review, it will be easier to suggest improvements on the published form

File: docs/getting-started/transactions/gosdk/usertx.md (hunk 3 above)
Comment (gennadylaventman) [reply docs/getting-started/transactions/gosdk/usertx.md:175]: Line 174 - two lines above change.

File: docs/getting-started/transactions/curl/datatx.md (hunk 23 above)
Comment (cendhu) [reply docs/getting-started/transactions/curl/datatx.md:213]: I will do this at once shot https://github.com/hyperledger-labs/orion-server/issues/284 It has been used in other places too

File: docs/getting-started/transactions/curl/datatx.md (hunk 24 above)
Comment (cendhu) [reply docs/getting-started/transactions/curl/datatx.md:497]: I will do this at once shot https://github.com/hyperledger-labs/orion-server/issues/284 It has been used in other places too

File: docs/getting-started/queries/gosdk/complex-data-query.md (hunk 16 above)
Comment (cendhu) [reply docs/getting-started/queries/gosdk/complex-data-query.md:7]: Done

File: docs/getting-started/queries/gosdk/complex-data-query.md (hunk 15 above)
Comment (cendhu) [reply docs/getting-started/queries/gosdk/complex-data-query.md:3]: Done. I never knew till now that Object is the correct word :-)

File: docs/getting-started/queries/gosdk/complex-data-query.md (hunk 17 above)
Comment (cendhu) [reply docs/getting-started/queries/gosdk/complex-data-query.md:36]: I did a search using `ack` to find all occurrences. I think `a data transaction` is fine.

File: docs/getting-started/queries/gosdk/simple-data-query.md (hunk 20 above)
Comment (cendhu) [reply docs/getting-started/queries/gosdk/simple-data-query.md:6]: Done

File: docs/getting-started/queries/gosdk/simple-data-query.md (hunk 19 above)
Comment (cendhu) [reply docs/getting-started/queries/gosdk/simple-data-query.md:3]: Done

File: docs/getting-started/queries/gosdk/db.md (hunk 18 above)
Comment (cendhu) [reply docs/getting-started/queries/gosdk/db.md:6]: Done. I did a search using `ack` command to find all occurrences. Not sure whether commands work correctly though :-(

File: docs/getting-started/queries/gosdk/user.md (hunk 21 above)
Comment (cendhu) [reply docs/getting-started/queries/gosdk/user.md:4]: Done

File: docs/getting-started/queries/gosdk/user.md (hunk 22 above)
Comment (cendhu) [reply docs/getting-started/queries/gosdk/user.md:6]: Done

File: docs/getting-started/queries/gosdk/complex-data-query.md (hunk 17 above)
Comment (tock-ibm) [reply docs/getting-started/queries/gosdk/complex-data-query.md]: 'a data transaction' is fine because transaction is the subject and is singular.

**Review by tock-ibm — APPROVED** [review]

**Pull request merged**